source/base/index_set.cc

```diff
@@ -547,9 +546,63 @@ IndexSet::make_trilinos_map (const MPI_Comm &communicator,
                         );
     }
 }
+#endif
 
 
-#endif
+
+bool
+IndexSet::is_linear (const MPI_Comm &communicator) const
+{
+  // If there is only one process, the IndexSet is always linear
+  const unsigned int n_ranks = Utilities::MPI::n_mpi_processes(communicator);
+  bool is_linear = (n_ranks==1);
+  const bool all_contiguous = (Utilities::MPI::min (is_contiguous() ? 1 : 0, communicator) == 1);
+  // non_contiguous IndexSets can't be linear
+  if ((all_contiguous) && (!is_linear))
+    {
+      is_linear = true;
+      // we know that there is only one interval
+      types::global_dof_index local_dofs[2];
+      local_dofs[0] = (n_elements()>0) ? *(begin_intervals()->begin())
+                      : numbers::invalid_dof_index ;
+      local_dofs[1] = (n_elements()>0) ? begin_intervals()->last()
+                      : numbers::invalid_dof_index;
+
+      const unsigned int my_rank = Utilities::MPI::this_mpi_process(communicator);
+// first gather all information on process 0
+      const unsigned int gather_size = (my_rank==0)?2*n_ranks:1;
+      std::vector<types::global_dof_index> global_dofs(gather_size);
+
+      MPI_Gather(local_dofs, 2, DEAL_II_DOF_INDEX_MPI_TYPE,
```

Comment:
Not sure if the code is much simpler that way, but if all sets are contiguous and their lengths sum to ``size``, you would just need to check that the start indices are sorted. That way you also transfer half as much data. This is still a global computation, sadly...